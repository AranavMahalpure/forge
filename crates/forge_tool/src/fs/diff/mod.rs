mod apply;

pub use apply::{FSReplace, FSReplaceInput};