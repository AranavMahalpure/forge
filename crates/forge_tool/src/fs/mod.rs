mod file_info;
mod fs_find;
mod fs_list;
mod fs_read;
mod fs_replace;
mod fs_write;
mod syn;

pub use file_info::*;
pub use fs_find::*;
pub use fs_list::*;
pub use fs_read::*;
pub use fs_replace::*;
pub use fs_write::*;
