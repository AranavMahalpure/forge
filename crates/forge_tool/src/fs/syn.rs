use std::path::Path;

use tree_sitter::{Language, Parser};

/// Maps file extensions to their corresponding Tree-sitter language parsers.
///
/// This function takes a file extension as input and returns the appropriate
/// Tree-sitter language parser if supported.
///
/// # Arguments
/// * `ext` - The file extension to get a language parser for
///
/// # Returns
/// * `Some(Language)` - If the extension is supported
/// * `None` - If the extension is not supported
///
/// # Supported Languages
/// * Rust (.rs)
/// * JavaScript/TypeScript (.js, .jsx, .ts, .tsx)
/// * Python (.py)
pub fn extension(ext: &str) -> Option<Language> {
    match ext.to_lowercase().as_str() {
        "rs" => Some(tree_sitter_rust::LANGUAGE.into()),
        "js" | "jsx" | "ts" | "tsx" => Some(tree_sitter_javascript::LANGUAGE.into()),
        "py" => Some(tree_sitter_python::LANGUAGE.into()),
        _ => None,
    }
}

/// Validates source code content using Tree-sitter parsers.
///
/// This function attempts to parse the provided content using a Tree-sitter
/// parser appropriate for the file's extension. It checks for syntax errors in
/// the parsed abstract syntax tree.
///
/// # Arguments
/// * `path` - The path to the file being validated (used to determine language)
/// * `content` - The source code content to validate
///
/// # Returns
/// * `Ok(())` - If the content is valid for the given language
/// * `Err(String)` - If validation fails, contains error description
///
/// # Note
/// Files with unsupported extensions are considered valid and will return
/// Ok(()). Files with no extension will return an error.
pub fn validate(path: impl AsRef<Path>, content: &str) -> Result<(), String> {
    let path = path.as_ref();

    // Get file extension
    let ext = path
        .extension()
        .and_then(|e| e.to_str())
        .ok_or_else(|| "File has no extension".to_string())?;

    // Get language for the extension
    let language = if let Some(lang) = extension(ext) {
        lang
    } else {
        // If we don't support the language, consider it valid
        return Ok(());
    };

    // Initialize parser
    let mut parser = Parser::new();
    parser.set_language(&language).map_err(|e| e.to_string())?;

    // Try parsing the content
    let tree = parser
        .parse(content, None)
        .ok_or_else(|| format!("Failed to parse {} as {}", path.display(), ext))?;

    // Find syntax errors in the tree
    let root_node = tree.root_node();
    if root_node.has_error() || root_node.is_error() {
        Err(format!(
            "Syntax error found in {} while parsing",
            path.display(),
        ))
    } else {
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::*;

    // Include language samples
    const RUST_VALID: &str = include_str!("lang/rust/valid.rs");
    const RUST_INVALID: &str = include_str!("lang/rust/invalid.rs");
    const JAVASCRIPT_VALID: &str = include_str!("lang/javascript/valid.js");
    const JAVASCRIPT_INVALID: &str = include_str!("lang/javascript/invalid.js");
    const PYTHON_VALID: &str = include_str!("lang/python/valid.py");
    const PYTHON_INVALID: &str = include_str!("lang/python/invalid.py");

    #[test]
    fn test_rust_valid() {
        let path = PathBuf::from("test.rs");
        assert!(validate(&path, RUST_VALID).is_ok());
    }

    #[test]
    fn test_rust_invalid() {
        let path = PathBuf::from("test.rs");
        assert!(validate(&path, RUST_INVALID).is_err());
    }

    #[test]
    fn test_javascript_valid() {
        let path = PathBuf::from("test.js");
        assert!(validate(&path, JAVASCRIPT_VALID).is_ok());
    }

    #[test]
    fn test_javascript_invalid() {
        let path = PathBuf::from("test.js");
        assert!(validate(&path, JAVASCRIPT_INVALID).is_err());
    }

    #[test]
    fn test_python_valid() {
        let path = PathBuf::from("test.py");
        assert!(validate(&path, PYTHON_VALID).is_ok());
    }

    #[test]
    fn test_python_invalid() {
        let path = PathBuf::from("test.py");
        assert!(validate(&path, PYTHON_INVALID).is_err());
    }

    #[test]
    fn test_unsupported_extension() {
        let content = "Some random content";
        let path = PathBuf::from("test.txt");
        assert!(validate(&path, content).is_ok());
    }

    #[test]
    fn test_no_extension() {
        let content = "Some random content";
        let path = PathBuf::from("test");
        assert!(validate(&path, content).is_err());
    }
}
