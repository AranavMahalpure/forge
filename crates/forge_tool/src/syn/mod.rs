mod validate;

pub use validate::{validate, Error};