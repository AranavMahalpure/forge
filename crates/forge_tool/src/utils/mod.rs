mod normalize_path;
mod path_validation;

#[cfg(test)]
pub use normalize_path::*;
pub use path_validation::*;
