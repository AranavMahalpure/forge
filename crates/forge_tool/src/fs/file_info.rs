use anyhow::Context;
use forge_domain::{NamedTool, ToolCallService, ToolDescription, ToolName};
use forge_tool_macros::ToolDescription;
use schemars::JsonSchema;
use serde::Deserialize;

#[derive(Deserialize, JsonSchema)]
pub struct FSFileInfoInput {
    /// The path of the file or directory to inspect, always provide the
    /// absolute path.
    pub path: String,
}

/// Request to retrieve detailed metadata about a file or directory at the
/// specified path. Returns comprehensive information including size, creation
/// time, last modified time, permissions, and type. Use this when you need to
/// understand file characteristics without reading the actual content.
#[derive(ToolDescription)]
pub struct FSFileInfo;

impl NamedTool for FSFileInfo {
    fn tool_name(&self) -> ToolName {
        ToolName::new("tool_forge_fs_info")
    }
}

#[async_trait::async_trait]
impl ToolCallService for FSFileInfo {
    type Input = FSFileInfoInput;

    async fn call(&self, input: Self::Input) -> Result<String, String> {
        let path = input.path.clone();
        let meta = tokio::fs::metadata(&path)
            .await
            .with_context(|| format!("Failed to get metadata for '{}'", path))
            .map_err(|e| e.to_string())?;
        Ok(format!("{:?}", meta))
    }
}

#[cfg(test)]
mod test {
    use tempfile::TempDir;
    use tokio::fs;

    use super::*;

    #[tokio::test]
    async fn test_fs_file_info_on_file() {
        let temp_dir = TempDir::new().unwrap();
        let file_path = temp_dir.path().join("test.txt");
        fs::write(&file_path, "test content").await.unwrap();

        let fs_info = FSFileInfo;
        let result = fs_info
            .call(FSFileInfoInput { path: file_path.to_string_lossy().to_string() })
            .await
            .unwrap();

        assert!(result.contains("FileType"));
        assert!(result.contains("permissions"));
        assert!(result.contains("modified"));
    }

    #[tokio::test]
    async fn test_fs_file_info_on_directory() {
        let temp_dir = TempDir::new().unwrap();
        let dir_path = temp_dir.path().join("test_dir");
        fs::create_dir(&dir_path).await.unwrap();

        let fs_info = FSFileInfo;
        let result = fs_info
            .call(FSFileInfoInput { path: dir_path.to_string_lossy().to_string() })
            .await
            .unwrap();

        assert!(result.contains("FileType"));
        assert!(result.contains("permissions"));
        assert!(result.contains("modified"));
    }

    #[tokio::test]
    async fn test_fs_file_info_nonexistent() {
        let temp_dir = TempDir::new().unwrap();
        let nonexistent_path = temp_dir.path().join("nonexistent");

        let fs_info = FSFileInfo;
        let result = fs_info
            .call(FSFileInfoInput { path: nonexistent_path.to_string_lossy().to_string() })
            .await;

        assert!(result.is_err());
    }
}
