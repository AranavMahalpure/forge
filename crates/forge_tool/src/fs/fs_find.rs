use std::collections::HashSet;
use std::path::Path;

use forge_domain::{Description, ToolCallService};
use forge_tool_macros::Description;
use forge_walker::Walker;
use schemars::JsonSchema;
use serde::Deserialize;

#[derive(Deserialize, JsonSchema)]
pub struct FSSearchInput {
    /// The path of the directory to search in (relative to the current working
    /// directory). This directory will be recursively searched.
    pub path: String,
    /// The regular expression pattern to search for. Uses Rust regex syntax.
    pub regex: String,
    /// Glob pattern to filter files (e.g., '*.ts' for TypeScript files). If not
    /// provided, it will search all files (*).
    pub file_pattern: Option<String>,
}

/// Request to perform a regex search across files in a specified directory,
/// providing context-rich results. This tool searches for patterns or specific
/// content across multiple files, displaying each match with encapsulating
/// context.
#[derive(Description)]
pub struct FSSearch;

#[async_trait::async_trait]
impl ToolCallService for FSSearch {
    type Input = FSSearchInput;
    type Output = Vec<String>;

    async fn call(&self, input: Self::Input) -> Result<Self::Output, String> {
        use regex::Regex;

        let dir = Path::new(&input.path);
        if !dir.exists() {
            return Err(format!("Directory '{}' does not exist", input.path));
        }

        // Create regex pattern - case-insensitive by default
        let pattern = format!("(?i){}", input.regex);
        let regex = Regex::new(&pattern).map_err(|e| format!("Invalid regex pattern: {}", e))?;

        let walker = Walker::new(dir.to_path_buf());
        let files = walker
            .get()
            .await
            .map_err(|e| format!("Failed to walk directory: {}", e))?;

        let mut matches = Vec::new();
        let mut seen_paths = HashSet::new();

        for file in files {
            if file.is_dir {
                continue;
            }

            let path = Path::new(&file.path);

            // Apply file pattern filter if provided
            if let Some(ref pattern) = input.file_pattern {
                let glob = glob::Pattern::new(pattern)
                    .map_err(|e| format!("Invalid glob pattern: {}", e))?;
                if let Some(filename) = path.file_name().and_then(|n| n.to_str()) {
                    if !glob.matches(filename) {
                        continue;
                    }
                }
            }

            // Skip if we've already processed this file
            if !seen_paths.insert(path.to_path_buf()) {
                continue;
            }

            // Full path for reading the file
            let full_path = dir.join(path);

            // Try to read the file content
            let content = match tokio::fs::read_to_string(&full_path).await {
                Ok(content) => content,
                Err(e) => {
                    // Skip binary or unreadable files silently
                    if e.kind() != std::io::ErrorKind::InvalidData {
                        matches.push(format!("Error reading {:?}: {}", path, e));
                    }
                    continue;
                }
            };

            // Process the file line by line
            for (line_num, line) in content.lines().enumerate() {
                if regex.is_match(line) {
                    // Format match in ripgrep style: filepath:line_num:content
                    matches.push(format!("{}:{}:{}", full_path.display(), line_num + 1, line));
                }
            }
        }

        if matches.is_empty() {
            Ok(vec![format!(
                "No matches found for pattern '{}' in path '{}'",
                input.regex, input.path
            )])
        } else {
            Ok(matches)
        }
    }
}

#[cfg(test)]
mod test {
    use tempfile::TempDir;
    use tokio::fs;

    use super::*;

    #[tokio::test]
    async fn test_fs_search_content() {
        let temp_dir = TempDir::new().unwrap();

        fs::write(temp_dir.path().join("test1.txt"), "Hello test world")
            .await
            .unwrap();
        fs::write(temp_dir.path().join("test2.txt"), "Another test case")
            .await
            .unwrap();
        fs::write(temp_dir.path().join("other.txt"), "No match here")
            .await
            .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "test".to_string(),
                file_pattern: None,
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 2);
        assert!(result.iter().any(|p| p.contains("test1.txt")));
        assert!(result.iter().any(|p| p.contains("test2.txt")));
    }

    #[tokio::test]
    async fn test_fs_search_with_pattern() {
        let temp_dir = TempDir::new().unwrap();

        fs::write(temp_dir.path().join("test1.txt"), "Hello test world")
            .await
            .unwrap();
        fs::write(temp_dir.path().join("test2.rs"), "fn test() {}")
            .await
            .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "test".to_string(),
                file_pattern: Some("*.rs".to_string()),
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 1);
        assert!(result.iter().any(|p| p.contains("test2.rs")));
    }

    #[tokio::test]
    async fn test_fs_search_with_context() {
        let temp_dir = TempDir::new().unwrap();
        let content = "line 1\nline 2\ntest line\nline 4\nline 5";

        fs::write(temp_dir.path().join("test.txt"), content)
            .await
            .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "test".to_string(),
                file_pattern: None,
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 1);
        assert!(result[0].contains("test line"));
    }

    #[tokio::test]
    async fn test_fs_search_recursive() {
        let temp_dir = TempDir::new().unwrap();

        let sub_dir = temp_dir.path().join("subdir");
        fs::create_dir(&sub_dir).await.unwrap();

        fs::write(temp_dir.path().join("test1.txt"), "test content")
            .await
            .unwrap();
        fs::write(sub_dir.join("test2.txt"), "more test content")
            .await
            .unwrap();
        fs::write(sub_dir.join("best.txt"), "this is proper\n test content")
            .await
            .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "test".to_string(),
                file_pattern: None,
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 3);
        assert!(result.iter().any(|p| p.contains("test1.txt")));
        assert!(result.iter().any(|p| p.contains("test2.txt")));
        assert!(result.iter().any(|p| p.contains("best.txt")));
    }

    #[tokio::test]
    async fn test_fs_search_case_insensitive() {
        let temp_dir = TempDir::new().unwrap();

        fs::write(
            temp_dir.path().join("test.txt"),
            "TEST CONTENT\ntest content",
        )
        .await
        .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "test".to_string(),
                file_pattern: None,
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 2);
        assert!(result.iter().any(|p| p.contains("TEST CONTENT")));
        assert!(result.iter().any(|p| p.contains("test content")));
    }

    #[tokio::test]
    async fn test_fs_search_no_matches() {
        let temp_dir = TempDir::new().unwrap();

        fs::write(temp_dir.path().join("test.txt"), "content")
            .await
            .unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "nonexistent".to_string(),
                file_pattern: None,
            })
            .await
            .unwrap();

        assert_eq!(result.len(), 1);
        assert!(result[0].contains("No matches found"));
    }

    #[tokio::test]
    async fn test_fs_search_invalid_regex() {
        let temp_dir = TempDir::new().unwrap();

        let fs_search = FSSearch;
        let result = fs_search
            .call(FSSearchInput {
                path: temp_dir.path().to_string_lossy().to_string(),
                regex: "[invalid".to_string(),
                file_pattern: None,
            })
            .await;

        assert!(result.is_err());
        assert!(result.unwrap_err().contains("Invalid regex pattern"));
    }
}
