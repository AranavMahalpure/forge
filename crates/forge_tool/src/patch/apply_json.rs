use std::path::Path;

use dissimilar::Chunk;
use forge_domain::{NamedTool, ToolCallService, ToolDescription, ToolName};
use forge_tool_macros::ToolDescription;
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use thiserror::Error;
use tokio::fs;

use crate::syn;
use crate::utils::assert_absolute_path;

/// Threshold for fuzzy matching. A score above this value is considered a
/// match. The score is calculated as the ratio of matching characters to total
/// characters.
const MATCH_THRESHOLD: f64 = 0.7;

/// Represents a potential patch match in the source text
#[derive(Debug)]
struct PatchMatch {
    text: String,
    similarity: f64,
}

impl PatchMatch {
    fn is_good_match(&self) -> bool {
        self.similarity >= MATCH_THRESHOLD
    }
}

#[derive(Debug, Error)]
enum Error {
    #[error("Failed to read/write file: {0}")]
    FileOperation(#[from] std::io::Error),
    #[error("Could not find match for search text: {0}")]
    NoMatch(String),
}

/// Find the best matching section using fuzzy matching
fn find_best_match(content: &str, search: &str) -> Option<PatchMatch> {
    // For exact matches, return immediately
    if content.contains(search) {
        return Some(PatchMatch { text: search.to_string(), similarity: 1.0 });
    }

    // Otherwise use fuzzy matching
    dissimilar::diff(content, search)
        .iter()
        .filter_map(|chunk| match chunk {
            Chunk::Equal(text) => {
                // Weight longer matches higher than shorter ones
                Some(PatchMatch {
                    text: text.to_string(),
                    similarity: text.to_string().chars().count() as f64
                        / search.chars().count() as f64,
                })
            }
            _ => None,
        })
        .filter(PatchMatch::is_good_match)
        .max_by(|a, b| {
            a.similarity
                .partial_cmp(&b.similarity)
                .unwrap_or(std::cmp::Ordering::Equal)
        })
}

fn apply_replacements(content: String, replacements: Vec<Replacement>) -> Result<String, Error> {
    // Iterate over all replacements and apply them one by one
    replacements
        .iter()
        .try_fold(content, |source, replacement| {
            if replacement.search.is_empty() {
                // Append mode - add content at the end
                Ok(format!("{}{}", source, replacement.content))
            } else {
                let patch = find_best_match(&source, &replacement.search)
                    .ok_or_else(|| Error::NoMatch(replacement.search.clone()))?;

                Ok(if replacement.content.is_empty() {
                    // Delete mode - remove the matched content
                    source.replacen(&patch.text, "", 1)
                } else {
                    // Replace mode - substitute matched content with new content
                    source.replacen(&patch.text, &replacement.content, 1)
                })
            }
        })
}

/// A single search and replace operation
#[derive(Deserialize, Serialize, JsonSchema, Debug, Clone)]
pub struct Replacement {
    pub search: String,
    pub content: String,
}

#[derive(Deserialize, JsonSchema)]
pub struct ApplyPatchJsonInput {
    pub path: String,
    pub replacements: Vec<Replacement>,
}

/// Finds and replaces all occurrences of the search text with the replacement
/// text in the file at the given path.
#[derive(ToolDescription)]
pub struct ApplyPatchJson;

impl NamedTool for ApplyPatchJson {
    fn tool_name() -> ToolName {
        ToolName::new("tool_forge_patch_v2")
    }
}

/// Format the modified content as XML with optional syntax warning
fn format_output(path: &str, content: &str, warning: Option<&str>) -> String {
    if let Some(w) = warning {
        format!(
            "<file_content\n  path=\"{}\"\n  syntax_checker_warning=\"{}\">\n{}</file_content>\n",
            path, w, content
        )
    } else {
        format!(
            "<file_content path=\"{}\">\n{}\n</file_content>\n",
            path,
            content.trim_end()
        )
    }
}

/// Process the file modifications and return the formatted output
async fn process_file_modifications(
    path: &Path,
    replacements: Vec<Replacement>,
) -> Result<String, Error> {
    let content = fs::read_to_string(path).await?;
    let content = apply_replacements(content, replacements)?;
    fs::write(path, &content).await?;

    let warning = syn::validate(path, &content).map(|e| e.to_string());
    Ok(format_output(
        path.to_string_lossy().as_ref(),
        &content,
        warning.as_deref(),
    ))
}

#[async_trait::async_trait]
impl ToolCallService for ApplyPatchJson {
    type Input = ApplyPatchJsonInput;

    async fn call(&self, input: Self::Input) -> Result<String, String> {
        let path = Path::new(&input.path);
        assert_absolute_path(path)?;

        process_file_modifications(path, input.replacements)
            .await
            .map_err(|e| e.to_string())
    }
}

#[cfg(test)]
mod test {
    use std::fmt::{self, Display};

    use super::*;

    // Test helpers
    #[derive(Debug)]
    struct PatchTest {
        initial: String,
        replacements: Vec<Replacement>,
        result: Option<String>,
    }

    impl fmt::Display for PatchTest {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            let replacements = self
                .replacements
                .iter()
                .map(|a| a.to_string())
                .fold("".to_string(), |a, b| format!("{a}{b}"));
            write!(
                f,
                "\n<!-- INITIAL -->{}\n<!-- REPLACEMENTS -->{}\n<!-- FINAL -->{}",
                self.initial,
                replacements,
                self.result
                    .as_ref()
                    .expect("Test must be executed before display")
            )
        }
    }

    impl PatchTest {
        fn new(initial: impl ToString) -> Self {
            PatchTest {
                initial: initial.to_string(),
                replacements: Default::default(),
                result: Default::default(),
            }
        }

        fn replace(mut self, search: impl ToString, replace: impl ToString) -> Self {
            self.replacements.push(Replacement::new(search, replace));
            self
        }

        // TODO: tests don't need to write files to disk
        fn execute(mut self) -> Result<Self, Error> {
            self.result = Some(apply_replacements(
                self.initial.clone(),
                self.replacements.clone(),
            )?);

            Ok(self)
        }
    }

    impl Display for Replacement {
        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
            write!(
                f,
                "\n<!-- SEARCH -->{}\n<!-- REPLACE -->{}",
                self.search, self.content
            )
        }
    }

    impl Replacement {
        fn new(search: impl ToString, replace: impl ToString) -> Self {
            Replacement { search: search.to_string(), content: replace.to_string() }
        }
    }

    /*
     * Basic Operations
     * Tests basic functionality like exact matches, empty inputs, and simple
     * cases
     */
    #[test]
    fn exact_match_single_word() {
        let actual = PatchTest::new("Hello World")
            .replace("World", "Forge")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn fuzzy_match_with_extra_characters() {
        let actual = PatchTest::new("fooo bar")
            .replace("foo", "baz") // Should match despite extra 'o'
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn append_empty_search() {
        let actual = PatchTest::new("foo").replace("", " bar").execute().unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn delete_empty_replace() {
        let actual = PatchTest::new("foo bar baz")
            .replace("bar ", "")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Multiple Replacements
     * Tests behavior when multiple replacements are performed
     */
    #[test]
    fn different_replacements_in_sequence() {
        let actual = PatchTest::new("foo bar")
            .replace("foo", "baz")
            .replace("bar", "qux")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn replaces_only_first_match() {
        let actual = PatchTest::new("foo bar foo")
            .replace("foo", "baz")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn sequential_replacements_on_identical_text() {
        let actual = PatchTest::new("same same same")
            .replace("same", "different") // First occurrence
            .replace("same", "changed") // Second occurrence
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn sequential_non_overlapping_replacements() {
        let actual = PatchTest::new("abcdef")
            .replace("abc", "123")
            .replace("def", "456")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Fuzzy Matching Behavior
     * Tests the fuzzy matching algorithm and similarity thresholds
     */
    #[test]
    fn exact_threshold_match() {
        let actual = PatchTest::new("foox") // 3/4 = 0.75, just above MATCH_THRESHOLD
            .replace("foo", "bar")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn just_below_threshold() {
        let result = PatchTest::new("fox") // 2/3 ≈ 0.67, just below MATCH_THRESHOLD
            .replace("foo", "bar")
            .execute();
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Could not find match"));
    }

    #[test]
    fn fuzzy_match_with_prefix() {
        let actual = PatchTest::new("foox baz foo")
            .replace("afoo", "bar") // Should match "foox" despite "a" prefix
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn prefers_exact_matches_over_fuzzy() {
        let actual = PatchTest::new("foo foox foo")
            .replace("foo", "bar") // Should replace exact "foo" first
            .replace("foo", "baz") // Should replace second exact "foo"
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Unicode and Special Characters
     * Tests handling of non-ASCII text and special characters
     */
    #[test]
    fn unicode_characters() {
        let actual = PatchTest::new("Hello 世界")
            .replace("世界", "🌍")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn emoji_replacement() {
        let actual = PatchTest::new("Hi 👋, how are you?")
            .replace("👋", "👍")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn mixed_unicode_ascii() {
        let actual = PatchTest::new("Hello 世界 World")
            .replace("世界 World", "地球")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Whitespace Handling
     * Tests preservation of whitespace, indentation, and line endings
     */
    #[test]
    fn preserve_indentation() {
        let actual = PatchTest::new("    indented\n        more indented")
            .replace("indented", "text")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn tab_characters() {
        let actual = PatchTest::new("no_tab\thas_tab")
            .replace("has_tab", "replaced")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn multiple_line_endings() {
        let actual = PatchTest::new("line1\nline2\rline3\r\nline4")
            .replace("line2", "replaced")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Error Cases
     * Tests error handling and edge cases
     */
    #[test]
    fn nested_replacements() {
        let actual = PatchTest::new("outer inner outer")
            .replace("outer inner outer", "modified inner")
            .replace("inner", "content")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn overlapping_matches() {
        let actual = PatchTest::new("abcabcabc")
            .replace("abca", "1234")
            .replace("cabc", "5678")
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    /*
     * Complex Replacements
     * Tests complicated scenarios like nested and overlapping matches
     */

    /*
     * Error Cases
     * Tests error handling and validation
     */

    #[test]
    fn delete_single_line_only() {
        let actual = PatchTest::new("line1\nline2\nline1\nline3")
            .replace("line1\n", "") // Delete first occurrence of line1 and its newline
            .execute()
            .unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn empty_search_text() {
        let actual = PatchTest::new("").replace("", "foo").execute().unwrap();
        insta::assert_snapshot!(actual);
    }

    #[test]
    fn no_match_found() {
        let result = PatchTest::new("foo").replace("bar", "baz").execute();
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Could not find match"));
    }

    #[test]
    fn fuzzy_match_below_threshold() {
        let result = PatchTest::new("fo").replace("foo", "bar").execute();
        assert!(result.is_err());
        assert!(result
            .unwrap_err()
            .to_string()
            .contains("Could not find match"));
    }
}
