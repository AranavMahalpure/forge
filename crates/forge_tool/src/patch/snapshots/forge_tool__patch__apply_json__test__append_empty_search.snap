---
source: crates/forge_tool/src/patch/apply_json.rs
expression: actual
snapshot_kind: text
---
<!-- INITIAL -->foo
<!-- REPLACEMENTS -->
<!-- SEARCH -->
<!-- REPLACE --> bar
<!-- FINAL -->foo bar
